### ISSUE TITLE
Slow clip_grad_norm_ because of .item() calls when run on device

### ISSUE THREAD

**Issue by redknightlois** [pytorch/pytorch#31474]
## Issue description

While profiling a slow training process of a quite deep network I got the following results:
![image](https://user-images.githubusercontent.com/1045411/71187912-c96f3480-225e-11ea-85b7-3cca0f34fdad.png)
Given I saw the problem was the `.item()` call would copy the values from the GPU crossing the PCI, I went ahead and built a version that would optimize that.

I replaced:
```python
total_norm = 0.0
for p in parameters:
   param_norm = p.grad.data.norm(norm_type)
   total_norm += param_norm ** norm_type
```
with:
```python
# PERF: We ensure we are all in the same device to avoid the penalty of moving data back and forth.
total_norm = torch.sum(torch.stack([torch.norm(p.grad,norm_type) for p in parameters]) ** norm_type)
```

There I saw quite an improvement:
![image](https://user-images.githubusercontent.com/1045411/71188107-18b56500-225f-11ea-984c-2d37618c2729.png)

But, to my surprise performance was still very bad for a very boilerplate function and the problem is actually the call to `.norm()` function. Which is horribly slow. 

## System Info

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Microsoft Windows 10 Pro
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.5
[pip] numpydoc==0.9.1
[pip] torch==1.3.1
[pip] torchvision==0.4.2
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      245
[conda] mkl-service               2.3.0            py37hb782905_0
[conda] mkl_fft                   1.0.14           py37h14836fe_0
[conda] mkl_random                1.1.0            py37h675688f_0
[conda] pytorch                   1.3.1           py3.7_cuda101_cudnn7_0    pytorch
[conda] torchvision               0.4.2                py37_cu101    pytorch

**Comment by ngimel** [comment]
Your change to clip_grad_norm implementation sounds good, do you mind submitting a PR? 
It looks like (on Pascal hardware at least) norm performance is not what it should be, it peaks at 180 GB/s instead of 550 GB/s. We'll look into it. 
Here's my benchmarking script
https://gist.github.com/ngimel/cac41f2f92cd95e8bb6781457a2bbce8

**Comment by Enealor** [comment]
I found a partial solution that increases the performance on the GPU, but not the CPU. I used
```
total_norm = torch.norm(torch.cat([p.grad.data.flatten() for p in parameters]), norm_type) ** norm_type
```
I wrote a profile script to test how effective this change is on my local machine, which I attached below. The profiler keeps track of how long is spent computing the total gradient norm of a model using three different methods: the original for loop, the function given by redknightlois, and the version I wrote above.

I tested it on both my CPU and GPU. On my GPU, the version I wrote takes ~20% the time of the original computation and ~40% of the version using `torch.stack`. However, when tested on the CPU, it adds time over the original due to the call to `torch.cat`. (Apparently the speed of `cat` on the CPU is a known issue, see #18634.) I mention this as when the speed issue with `cat` is resolved, this version might be a better choice for computing the total norm.

[profile_total_norm.zip](https://github.com/pytorch/pytorch/files/3986163/profile_total_norm.zip)

**Comment by ngimel** [comment]
@enealor your solution also uses way more memory (essentially, you need to allocate an additional buffer big enough to hold all the parameters), whereas stack-based solution uses very little additional memory, and also likely does not degrade CPU performance, though I have not tested. So I'd prefer to go with original stack-based solution for now. Ideally, we should be able to operate on the lists of tensors as fast as on a single flattened tensor, that avoids problems with cat performance and with additional memory, and @mcarilli has signed up to make this a reality. 
@enealor FYI there's a PR out that improves cat performance on CPU #30806.

**Comment by ngimel** [comment]
#31903 improves performance of individual norm operations, but redefinition of clip_grad_norm
```
total_norm = torch.sum(torch.stack([torch.norm(p.grad,norm_type) for p in parameters]) ** norm_type)
```
is still needed.
